modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/CrossNavigationEnrichment.scala

```diff
@@ -0,0 +1,213 @@
+   * @param config The cross_navigation_config enrichment JSON
+   * @param schemaKey provided for the enrichment, must be supported by this enrichment
+   * @return a CrossNavigation configuration
+   */
+  override def parse(
+    config: Json,
+    schemaKey: SchemaKey,
+    localMode: Boolean = false
+  ): ValidatedNel[String, CrossNavigationConf] =
+    (for {
+      _ <- isParseable(config, schemaKey)
+    } yield CrossNavigationConf(schemaKey)).toValidatedNel
+
+  /**
+   * Wrapper around CU.decodeBase64Url.
+   * If passed an empty string returns Right(None).
+   * @param str The string to decode
+   * @return either the decoded string or enrichment failure
+   */
+  private def decodeWithFailure(str: String): Either[FailureDetails.EnrichmentFailure, Option[String]] =
+    CU.decodeBase64Url(str) match {
+      case Right(r) => Option(r).filter(_.trim.nonEmpty).asRight
+      case Left(msg) =>
+        FailureDetails
+          .EnrichmentFailure(
+            None,
+            FailureDetails.EnrichmentFailureMessage.Simple(msg)
+          )
+          .asLeft
+    }
+
+  /**
+   * Wrapper around EE.extractTimestamp
+   * If passed an empty string returns Right(None).
+   * @param str The string to extract the timestamp from
+   * @return either the extracted timestamp or enrichment failure
+   */
+  private def extractTstamp(str: String): Either[FailureDetails.EnrichmentFailure, Option[String]] =
+    str match {
+      case "" => None.asRight
+      case s => EE.extractTimestamp("sp_dtm", s).map(_.some)
+    }
+
+  /**
+   * Converts a timestamp to an ISO-8601 format
+   *
+   * @param tstamp The timestamp expected as output of EE.extractTimestamp
+   * @return ISO-8601 timestamp
+   */
+  private def reformatTstamp(tstamp: Option[String]): Option[String] = {
+    val pFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS")
+    val formatter = DateTimeFormatter.ISO_DATE_TIME
+
+    tstamp match {
+      case Some(t) =>
+        Some(formatter.format(pFormatter.parse(t)).replaceAll(" ", "T") + "Z")
+      case None => None
+    }
+  }
+
+  /**
+   * Finalizes the cross navigation map by reformatting its timestamp key
+   *
+   * @param inputMap A Map of cross navigation properties
+   * @return The finalized Map
+   */
+  private def finalizeCrossNavigationMap(inputMap: Map[String, Option[String]]): Map[String, Option[String]] =
+    inputMap
+      .map({
+        case ("timestamp", t) => ("timestamp" -> reformatTstamp(t))
+        case kvPair => kvPair
+      })
+
+  /**
+   * Parses the QueryString into a Map
+   * @param sp QueryString
+   * @return either a map of query string parameters or enrichment failure
+   */
+  def makeCrossDomainMap(sp: String): Either[FailureDetails.EnrichmentFailure, Map[String, Option[String]]] =
+    sp.split("\\.", -1)
+      .padTo(
+        CrossNavProps.size,
+        ""
+      )
+      .toList match {
+      case List(
+            duidElt,
```

Comment:
It seems like there is too much repetition here. I think we can reduce the repetition by putting transformation functions and props name together. I've implemented what had in mind [here](https://github.com/snowplow/enrich/tree/feat/cross_navigation_cleanup). Let me know if you see any problem with it.